File: week-6/04/count-game.cpp
string counterGame(long n) {
    bool ric = true;  // Initialize a boolean variable to keep track of Richard's turn
    for (long m = n - 1; m != 0; m >>= 1) {  // Iterate through each binary digit of n except the most significant one
        if ((m & 1) == 1) ric = !ric;  // If the current binary digit is 1, toggle the turn
    }
    return ric ? "Richard" : "Louise";  // Return the winner based on the final turn
}

File: week-6/04/count-game.py
def counterGame(n):
    # Convert the integer n to its binary representation and remove the '0b' prefix
    binary = bin(n)[2:]
    
    # Count the number of '1's in the binary representation
    ones = len([i for i in binary if i == "1"])
    
    # Count the number of trailing '0's in the binary representation
    right_zeros = len(binary) - len(binary.rstrip("0"))
    
    # Determine the winner based on the number of '1's and trailing '0's
    return "Louise" if (ones + right_zeros - 1) % 2 else "Richard"
